src/pydicom/pixels/utils.py

```diff
@@ -2098,3 +2170,246 @@ def unpack_bits(src: bytes, as_array: bool = True) -> "np.ndarray | bytes":
+        np.bitwise_or(shifted[: len(ms_bytes)], ms_bytes, out=shifted[: len(ms_bytes)])
+
+        # Mask out any bits at the end that are not in this frame
+        np.bitwise_and(
+            shifted[-1:],
+            np.frombuffer(final_byte_mask, np.uint8),
+            out=shifted[-1:],
+        )
+
+        out = shifted.tobytes()
+
+    else:
+        # Start pixel is in the middle of a byte so need to shift bits between
+        # bytes. Without NumPy we fall back to simply unpacking all bytes
+        # containing pixels from this frame and repacking the relevant pixels
+        frame_bytes = src[byte_start:byte_end]
+        unpacked = unpack_bits(frame_bytes, as_array=False)
+
+        unpacked_frame = unpacked[pixel_start % 8 : pixel_start % 8 + frame_length]
+        out = pack_bits(unpacked_frame, pad=pad)
+
+    # Pad to multiple of 2 bytes
+    if pad and len(out) % 2 == 1:
+        out = b"".join([out, b"\x00"])
+
+    return out
+
+
+def concatenate_packed_frames(
+    frames: Sequence[bytes | bytearray],
+    frame_length: int,
+    pad: bool = True,
+) -> bytes:
+    """Concatenate individual bit-packed frames.
+
+    This should only be used for single bit data (Bits Allocated = 1) in a
+    "bit-packed" representation (one byte containing 8 pixels). Concatenating
+    such frames may require combining pixels from multiple frames within a
+    single byte.
+
+    Parameters
+    ----------
+    frames : Sequence[bytes | bytearray]
+        List of individual frames in native bit-packed representation.
+    frame_length : int
+        Number of pixels in each frame.
+    pad : bool
+        Whether to zero-pad the resulting bytes to a multiple of 2 bytes.
+
+    Returns
+    -------
+    bytes
+        Bit-packed representation of all frames combined.
+
+    Note
+    ----
+    In certain situations, this function will run faster and use less memory if
+    NumPy is installed. Otherwise, it will fall back to a slower method using
+    native Python functions.
+    """
+    if frame_length % 8 == 0:
+        # Simple concatenation since frame boundaries occur on byte boundaries
+        out = b"".join(frames)
+    else:
+        if HAVE_NP:
+            # Efficient implementation using NumPy to shift bits between bytes
+
+            # Pre-allocate an output array
+            total_bytes = math.ceil(len(frames) * frame_length / 8)
+            bytes_per_frame = math.ceil(frame_length / 8)
+            if pad and total_bytes % 2 == 1:
+                total_bytes += 1
+            out_arr = np.zeros(total_bytes, dtype=np.uint8)
+
+            # Temporary buffer re-used in the loop
+            tmp = np.empty(bytes_per_frame, dtype=np.uint8)
+
+            for i, frame in enumerate(frames):
+                pixel_start = i * frame_length
+                bit_offset = pixel_start % 8
+
+                # Index of the byte containing the first pixel of this frame
+                b = math.floor(pixel_start / 8)
```

Comment:
`b` isn't a very descriptive variable name

Reply:
Renamed to `first_byte_index` in [e9ea3de](https://github.com/pydicom/pydicom/pull/2216/commits/e9ea3deb538f6786f5a80e0e826d7aaf7a40678c)